FastSurferCNN/segstats.py

```diff
@@ -0,0 +1,726 @@
+              "normRange": maxes.get(lab, 0.) - mins.get(lab, 0.)} for lab in labels]
+    if return_maps:
+        return table, {"nbr": full_nbr_label, "segmean": full_seg_mean, "nbrmean": full_nbr_mean, "pv": full_pv,
+                       "ipv": full_ipv}
+    return table
+
+
+def global_stats(lab: _IntType, norm: npt.NDArray[_NumberType], seg: npt.NDArray[_IntType],
+                 out: Optional[npt.NDArray[bool]] = None, robust_percentage: Optional[float] = None) \
+        -> Union[Tuple[_IntType, int],
+                 Tuple[_IntType, int, int, _NumberType, _NumberType, float, float, float, npt.NDArray[bool]]]:
+    """Computes Label, Number of voxels, 'robust' number of voxels, norm minimum, maximum, sum, sum of squares and
+    6-connected border of label lab."""
+    bin_array = cast(npt.NDArray[bool], seg == lab)
+    data = norm[bin_array].astype(int if np.issubdtype(norm.dtype, np.integer) else float)
+    nvoxels: int = data.shape[0]
+    # if lab is not in the image at all
+    if nvoxels == 0:
+        return lab, 0
+    # compute/update the border
+    if out is None:
+        out = seg_borders(bin_array, True, cmp_dtype="int8").astype(bool)
+    else:
+        out[:] = seg_borders(bin_array, True, cmp_dtype="int").astype(bool)
+
+    if robust_percentage is not None:
+        data = np.sort(data)
+        sym_drop_samples = int((1 - robust_percentage / 2) * nvoxels)
+        data = data[sym_drop_samples:-sym_drop_samples]
+        _min: _NumberType = data[0].item()
+        _max: _NumberType = data[-1].item()
+        __voxel_count = nvoxels - 2 * sym_drop_samples
+    else:
+        _min = data.min().item()
+        _max = data.max().item()
+        __voxel_count = nvoxels
+    _sum: float = data.sum().item()
+    sum_2: float = (data * data).sum().item()
+    # this is independent of the robustness criterium
+    volume: float = np.sum(np.logical_and(bin_array, np.logical_not(out))).astype(float).item()
+    return lab, nvoxels, __voxel_count, _min, _max, _sum, sum_2, volume, out
+
+
+def patch_filter(pos: Tuple[int, int, int], mask: npt.NDArray[bool],
+                 global_crop: Tuple[slice, ...], patch_size: int = 32) \
+        -> Tuple[bool, Sequence[slice]]:
+    """Returns, whether there are mask-True voxels in the patch starting at pos with size patch_size and the resulting
+    patch shrunk to mask-True regions."""
+    # create slices for current patch context (constrained by the global_crop)
+    patch = [slice(p, min(p + patch_size, slice_.stop)) for p, slice_ in zip(pos, global_crop)]
+    # crop patch context to the image content
+    return crop_patch_to_mask(mask, sub_patch=patch)
+
+
+def crop_patch_to_mask(mask: npt.NDArray[_NumberType],
+                       sub_patch: Optional[Sequence[slice]] = None) \
+        -> Tuple[bool, Sequence[slice]]:
+    """Crop the patch to regions of the patch that are non-zero. Assumes mask is always positive. Returns whether there
```

Comment:
"Crop the patch to regions of the mask that are non-zero..."

Reply:
Yep, that is basically inherited from prior versions. I will update this docstring.